cmds/core/md5sum/md5sum.go

```diff
@@ -7,84 +7,57 @@ package main
 
 import (
 	"crypto/md5"
+	"flag"
 	"fmt"
 	"io"
 	"log"
 	"os"
 
-	"github.com/spf13/pflag"
+	"github.com/u-root/u-root/pkg/uroot/util"
 )
 
-func getInput() (input []byte, err error) {
-	return io.ReadAll(os.Stdin)
-}
-
-func helpPrinter() {
-	fmt.Printf("Usage:\nmd5sum <File Name>\n")
-	pflag.PrintDefaults()
-	os.Exit(0)
-}
-
-func versionPrinter() {
-	fmt.Println("md5sum utility, URoot Version.")
-	os.Exit(0)
-}
-
-func calculateMd5Sum(fileName string, data []byte) string {
-	if len(data) > 0 {
-		return fmt.Sprintf("%x", md5.Sum(data))
-	}
-
-	fileDesc, err := os.Open(fileName)
-	if err != nil {
-		log.Fatal(err)
-	}
-	defer fileDesc.Close()
-
-	md5Generator := md5.New()
-	if _, err := io.Copy(md5Generator, fileDesc); err != nil {
-		log.Fatal(err)
-	}
-
-	md5Sum := fmt.Sprintf("%x", md5Generator.Sum(nil))
-	return md5Sum
-}
+var (
+	help = flag.Bool("h", false, "Show this help and exit")
+)
 
-func main() {
-	var (
-		help    bool
-		version bool
-		input   []byte
-		err     error
-	)
-	cliArgs := ""
-	pflag.BoolVarP(&help, "help", "h", false, "Show this help and exit")
-	pflag.BoolVarP(&version, "version", "v", false, "Print Version")
-	pflag.Parse()
+var usage = "md5sum: md5sum <File Name>"
 
-	if help {
-		helpPrinter()
+func md5Sum(w io.Writer, r io.Reader, args ...string) (err error) {
+	if *help {
+		util.Usage(usage)
+		flag.Usage()
+		return nil
 	}
 
-	if version {
-		versionPrinter()
+	fileName := ""
+	if len(args) >= 1 {
+		fileName = args[0]
 	}
-
-	if len(os.Args) >= 2 {
-		cliArgs = os.Args[1]
-	}
-	if cliArgs == "" {
-		input, err = getInput()
+	if fileName == "" {
+		input, err := io.ReadAll(r)
```

Comment:
Again, you can not count on ReadAll, because on small devices, the stream can be larger than memory. I've had this happen. I don't think it will take much more to fix this, it's very close. The actual md5sum code is so compact (4 lines) that you might as well just check if len(args) == 0 and do those 4 lines. 

Of course, I still find this construct a bit awkward: a function called md5sum contains a call to help? The io.Reader is an optionally used argument if len(args) is 0? I think this could still use a bit more shaping. I think there are more like 3 functions in here than one. We need to be careful that our coverage goals don't push us to writing weird code :-)

Reply:
I refactored it again. I removed the help flag because it was useless and the flag pkg does support -h by default. I hope now you are satisfied. If not maybe you can give more concrete informations, what exactly you imagine.